Sources/AWSSDKSwiftCore/AWSClient.swift

```diff
@@ -234,107 +194,170 @@ extension AWSClient {
+        operation operationName: String,
+        path: String,
+        httpMethod: String,
+        input: Input,
+        with configuration: ServiceConfig,
+        on eventLoop: EventLoop? = nil) -> EventLoopFuture<Output>
+    {
+        self.execute(on: eventLoop, configuration: configuration, operationName: operationName) {
+            try AWSRequest(
+                operation: operationName,
+                path: path,
+                httpMethod: httpMethod,
+                input: input,
+                configuration: configuration)
+        }.recordMetrics(for: configuration.service, operation: operationName)
+    }
+    
+    private func execute(on eventLoop: EventLoop?, configuration: ServiceConfig, _ createRequest: @escaping () throws -> AWSRequest)
+        -> EventLoopFuture<AWSHTTPResponse>
+    {
         let eventLoop = eventLoop ?? eventLoopGroup.next()
-        let future: EventLoopFuture<Output> = credentialProvider.getCredential(on: eventLoop).flatMapThrowing { credential in
-            let signer = AWSSigner(credentials: credential, name: self.serviceConfig.signingName, region: self.serviceConfig.region.rawValue)
-            let awsRequest = try self.createAWSRequest(
-                        operation: operationName,
-                        path: path,
-                        httpMethod: httpMethod,
-                        input: input)
-            return awsRequest.createHTTPRequest(signer: signer)
-        }.flatMap { request in
-            return self.invoke(request, on: eventLoop)
-        }.flatMapThrowing { response in
-            return try self.validate(operation: operationName, response: response)
+        return credentialProvider.getCredential(on: eventLoop)
+            .flatMapThrowing { credential in
+                // create the request and apply middleware
+                let request = try createRequest().applyMiddlewares(configuration.middlewares + self.middlewares)
+                // create signer and sign request
+                let signer = AWSSigner(
+                    credentials: credential,
+                    name: configuration.signingName,
+                    region: configuration.region.rawValue)
+                return request.createHTTPRequest(signer: signer)
+            }
+            .flatMap { request in
+                // execute request
+                return self.invoke(request, on: eventLoop)
+            }
+            .flatMapErrorThrowing { (error) -> AWSHTTPResponse in
+                // transform the error into an AWSError if possible
+                guard case AWSClient.InternalError.httpResponseError(let response) = error else {
+                    throw error
+                }
+                throw Self.createError(for: response, configuration: configuration)
+            }
+    }
+    
+    private func execute(on eventLoop: EventLoop?, configuration: ServiceConfig, _ createRequest: @escaping () throws -> AWSRequest)
```

Comment:
While the restructure of the send/execute functions ensures we aren't repeating code the code path is a little confusing now. I had to run the code to see how each of the functions were called. Can we add some function headers defining what these functions do individually